### DISCUSSION TITLE
Tile with space between images + global border

### DISCUSSION THREAD

**Discussion by tergan1** [ImageMagick/ImageMagick#2803] [Help]
Hello.
How to tile a rose with 5px space between roses and 15px border all over the image ?
[https://i.imgur.com/0Dp0F3J.png](url)
![alt text](https://i.imgur.com/0Dp0F3J.png)
Can it be created by two solutions? 
First solution With "tile-offset +5+5" and Second solution with "-geometry +5+5" ?

Now I have this:
First (need to add tile-offset 5px and border 15px)
`montage rose: rose: rose: rose: -tile 2x -mode Concatenate !output1.png`
Second (need to add space 5px and border 15px)
`montage rose: rose: rose: rose: -tile 2x -geometry +0+0 !output2.png`

**Comment by tergan1** [comment]
Here is one solution but 4px space (not 5px)
`montage rose: rose: rose: rose: -background magenta -tile 2x -geometry +2+2 miff:- | convert - -bordercolor magenta -trim -border 15 !output022.png`

**Comment by fmw42** [comment]
This should do it - get a 5 pixel spacing between images and 15 pixel border. Put a 2 pixel border around each image. Then montage with gap of 1 pixel (-geometry +1+1). The put a 13 pixel border around the montage

`convert rose: -bordercolor magenta -border 2 -duplicate 3 miff:- | montage - -background magenta -geometry +1+1 -tile 2x2 miff:- | convert - -bordercolor magenta -border 13 x.png`

![x](https://user-images.githubusercontent.com/13458218/99008607-59634000-24fb-11eb-93e8-24fc87de0dd1.png)

**Reply by tergan1** [reply]
result is 6px between images and 16px border (not 5px and 15px)

**Reply by fmw42** [reply]
The thickness of the magenta is 5 between image and 15 around the outside. That is what you requested by your diagram. If you want it 4 and 14 so that the distance between image pixels is 5 and 15, then

`convert rose: -bordercolor magenta -border 2 -duplicate 3 miff:- | montage - -background magenta -geometry +0+0 -tile 2x2 miff:- | convert - -bordercolor magenta -border 13 x.png`

![x](https://user-images.githubusercontent.com/13458218/99140253-f05ff300-25f4-11eb-961b-c7496af26110.png)

**Reply by tergan1** [reply]
result is 4px between images and 15px border (not 5px and 15px)

**Reply by fmw42** [reply]
Please define the spacing in terms of the magenta thickness that you want.

**Reply by tergan1** [reply]
I already described the spacing in the start of this thread (the image with arrows). You can try code from "snibgo" - It is works perfectly

**Reply by fmw42** [reply]
It is not clear what spacing you want. Is it only the magenta color thickness or the distance from image to image. If snibgo's code works, then just use it. The number of magenta pixels is 4 between image (so 5 pixels from one image to the next) and 14 around the outside (so 15 pixels from border to any image).  Note when you measure, you have to include the first pixel. The thickness, for example from 85 to 88 is 4 pixels. That is the count of pixels is 4.

**Reply by tergan1** [reply]
Magenta color thickness = the distance from image to image.

Here is the pixel count between images, which I need:
![alt text](https://i.imgur.com/izPqcpH.png)

Here is the pixel count of border, which I need:
![alt text](https://i.imgur.com/1BzVxFt.png)

**Reply by fmw42** [reply]
I think you are looking at the wrong image. Try this one.

![99140253-f05ff300-25f4-11eb-961b-c7496af26110](https://user-images.githubusercontent.com/13458218/99153643-87ad6080-265e-11eb-88cf-e99413822e27.png)

**Reply by tergan1** [reply]
4px between images, need 5px like on the big image above
![alt text](https://i.imgur.com/vKXmEZ1.png)

**Reply by fmw42** [reply]
Try this:

```
magick rose: -bordercolor magenta -border 2 -duplicate 3 miff:- | magick montage - -background magenta -geometry +1+1 -tile 2x2 miff:- | magick  - -chop "1x1+%[fx:w/2]+%[fx:h/2]" miff:- | magick  - -bordercolor magenta -border 12 x.png
```

![x](https://user-images.githubusercontent.com/13458218/99161807-ad605700-26aa-11eb-81fd-6b125f8d9ca1.png)

See -chop at https://imagemagick.org/Usage/crop/#chop

**Reply by tergan1** [reply]
Bat file can't recognise this formula "1x1+%[fx:w/2]+%[fx:h/2]"
![alt text](https://i.imgur.com/ESQEHgu.png)

**Reply by snibgo** [reply]
In BAT scripts, double every %, so they are %%.

**Reply by tergan1** [reply]
Yes, it is a good result. 5px between images and 15 border.
Added "%%" and "rose:" for replace to "image1234.png"
`magick rose: rose: rose: rose: -bordercolor magenta -border 2 miff:- | magick montage - -background magenta -geometry +1+1 -tile 2x2 miff:- | magick  - -chop "1x1+%%[fx:w/2]+%%[fx:h/2]" miff:- | magick  - -bordercolor magenta -border 12 !output_6_1_4.png`

**Reply by fmw42** [reply]
This version only works for 2x2 grid of images, not 3x3 or 4x4 or any other

**Reply by fmw42** [reply]
Here is a better version that works for other number of cells. It should be more intuitive and explicitly uses your dimensions as specified.

- duplicated the input 3*3-1=8 more times for a grid of 3x3 or 9 cells
- extent each image by 5 pixels on the right and bottom
- montage them in 3x3 tiles with no spacing
- add a 15 pixel border all around
- use -extent to crop off the 5 pixels on the right and bottom

`magick rose: -duplicate "%[fx:3*3-1]" -background magenta -gravity northwest -extent "%[fx:w+5]x%[fx:h+5]" miff:- | magick montage - -geometry +0+0 -tile 3x3 miff:- | magick - -bordercolor magenta -border 15 miff:- | magick - -gravity northwest -extent "%[fx:w-5]x%[fx:h-5]" x.png
`

or in multi-line format that is easier to read (using Windows new line ^ characters)

```
magick rose: -duplicate "%[fx:3*3-1]" ^
-background magenta -gravity northwest -extent "%[fx:w+5]x%[fx:h+5]" miff:- | ^
magick montage - -geometry +0+0 -tile 3x3 miff:- | ^
magick - -bordercolor magenta -border 15 miff:- | ^
magick - -gravity northwest -extent "%[fx:w-5]x%[fx:h-5]" ^
x.png
```


![x](https://user-images.githubusercontent.com/13458218/99171434-e524ca00-26bd-11eb-8deb-3040b2ce7f00.png)

In Windows .bat, double the % to %%

**Reply by tergan1** [reply]
Perfect. Absolutely perfect.

**Comment by snibgo** [comment]
Or in a single command like this:
```
magick rose: -gravity NorthWest -background Magenta -splice 5x5 ( +clone ) +append ( +clone ) -append -splice 10x10 -gravity SouthEast -splice 15x15 x.png
```

**Reply by tergan1** [reply]
working, but so weird code, not noob friendly

**Reply by fmw42** [reply]
The first splice simply adds 5 pixels on the top and left, the copies the image and appends them horizontally. Then copies that and appends vertically. The next splice adds 10 on the top and left and the last one 15 on the right and bottom.

**Reply by snibgo** [reply]
When developing or debugging or understanding complex commands, I often use different colours to understand what is happening. Windows BAT syntax:
```
magick ^
  rose: ^
  -gravity NorthWest -background Magenta -splice 5x5 ( +clone ) +append ( +clone ) -append ^
  -background Yellow -splice 10x10 ^
  -gravity SouthEast -background Blue -splice 15x15 tileexamp.png
```
<img src="http://snibgo.com/imforums/tileexamp.png" />

**Reply by tergan1** [reply]
Nice imagination. How to add 4 different images? Without "+clone". Just "rose: rose: rose: rose:" (like in the first message of this topic) - so people can replace "rose:" on different "image1.png image2.png image3.png image4.png"

**Comment by snibgo** [comment]
```
magick ^
  ( rose: rose: +append ) ^
  ( rose: rose: +append ) ^
  -append ^
  -gravity NorthWest -background Magenta -splice 5x5 ^
  -background Yellow -splice 10x10 ^
  -gravity SouthEast -background Blue -splice 15x15 tileexamp.png
```
Change each "rose:" to any file. But they should be the same sizes.

**Reply by tergan1** [reply]
result is without space between images
![alt text](https://i.imgur.com/cB4IzPJ.png)

**Comment by snibgo** [comment]
So it is, sorry. Try:
```
magick ^
  -gravity NorthWest -background Magenta ^
  ( rose: rose: -splice 5x5 +append ) ^
  ( rose: rose: -splice 5x5 +append ) ^
  -append ^
  -background Yellow -splice 10x10 ^
  -gravity SouthEast -background Blue -splice 15x15 tileexamp.png
```

**Reply by tergan1** [reply]
Excellent !

![alt text](https://i.imgur.com/soAyvaQ.png)

**Reply by tergan1** [reply]
both answers are good, how to choose one of them so as not to offend anyone?

**Comment by snibgo** [comment]
Don't worry about it. We won't be offended.

**Comment by fmw42** [comment]
Take whichever you feel most comfortable about using. We won't be offended.
